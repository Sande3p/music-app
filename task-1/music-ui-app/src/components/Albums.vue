<template>
  <div class="card text-center m-3">
    <div class="search-bar">
      <input
        class="t-search"
        type="search"
        v-model="q"
        placeholder="Search album"
        v-on:keyup="searchAlbum"
        maxlength=256
      />
    </div>
    <p class="loading">{{ loading ? "Loading..." : "" }}</p>
    <div class="card-body">
      <div v-for="item in albumList" :key="item.collectionId">
        <AlbumTile :item="item" />
      </div>
    </div>
  </div>
</template>

<script>
import axios from "axios";
import AlbumTile from "./AlbumTile";
import config from "../../config/index";

export default {
  name: "alubms",
  components: {
    AlbumTile,
  },
  data() {
    return {
      fetchedAlbums: [],
      albumList: [],
      q: "",
      loading: true,
    };
  },
  methods: {
    searchAlbum(e) {
      this.albumList = [];
      this.fetchedAlbums.map((item) => {
        item.collectionName &&
          item.collectionName.toLowerCase().indexOf(this.q.toLowerCase()) >=
            0 &&
          this.albumList.push(item);
        return item;
      });
      this.q = e.target.value;
    },
  },
  watch: {
    q: function (newValue) {
      this.$emit("input", newValue);
    },
  },
  created() {
    // Simple GET request using axios
    this.loading = true;
    axios
      .get(config.API_URL + "/album?artist=taylor swift")
      .then((response) => {
        this.fetchedAlbums = response.data ? response.data.results : [];
        this.albumList = { ...this.fetchedAlbums };
        this.loading = false;
      })
      .catch((error) => {
        this.loading = false;
        console.log(error);
      });
  },
};
</script>
