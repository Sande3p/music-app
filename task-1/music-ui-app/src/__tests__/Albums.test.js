import { mount } from "@vue/test-utils";
import Albums from '../components/Albums.vue';

describe('Albums.test.js', () => {
  let cmp;

  beforeEach(async () => {
    cmp = mount(Albums)
    await cmp.setData({
      fetchedAlbums: [{
        "wrapperType": "collection",
        "collectionType": "Album",
        "artistId": 909253,
        "collectionId": 1469577723,
        "amgArtistId": 468749,
        "artistName": "Jack Johnson",
        "collectionName": "Demo album name",
        "collectionCensoredName": "Demo album name",
        "artworkUrl100": "https://is1-ssl.mzstatic.com/image/thumb/Music115/v4/ae/4c/d4/ae4cd42a-80a9-d950-16f5-36f01a9e1881/source/100x100bb.jpg"
      },
      {
        "wrapperType": "collection",
        "collectionType": "Album",
        "artistId": 909254,
        "collectionId": 1469577724,
        "amgArtistId": 468749,
        "artistName": "Jack Johnson",
        "collectionName": "New album name",
        "collectionCensoredName": "New album name",
        "artworkUrl100": "https://is1-ssl.mzstatic.com/image/thumb/Music115/v4/ae/4c/d4/ae4cd42a-80a9-d950-16f5-36f01a9e1881/source/100x100bb.jpg"
      }]
    })
  });

  it("has received 'fetchedAlbums' property", () => {
    expect(cmp.vm.fetchedAlbums.length).toEqual(2)
  })

  it("'albumList' is [] by default", () => {
    expect(cmp.vm.albumList).toEqual([])
  })

  it("expect seach string to have maxlength of 256", async () => {
    const searchText = cmp.find('input.t-search')
    expect(searchText.attributes('maxlength')).toBe('256'); 
  });

  it('Entering `new` in search input, set query string to `new` & search result to have 1 item ', async () => {

    const searchText = cmp.find('input.t-search')
    searchText.setValue('new')
    await searchText.trigger('keyup')
    // console.log(cmp.vm.albumList)
    expect(cmp.vm.q).toBe('new')
    expect(cmp.vm.albumList.length).toBe(1)
  })

  it("has the expected html structure", () => {
    expect(cmp.element).toMatchSnapshot();
  });


});