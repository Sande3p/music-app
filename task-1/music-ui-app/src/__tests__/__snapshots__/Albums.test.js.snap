// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Albums.test.js has the expected html structure 1`] = `
<div
  class="card text-center m-3"
>
  <div
    class="search-bar"
  >
    <input
      class="t-search"
      maxlength="256"
      placeholder="Search album"
      type="search"
    />
  </div>
   
  <p
    class="loading"
  >
    Loading...
  </p>
   
  <div
    class="card-body"
  />
</div>
`;
